enterprise/server/remote_execution/operation/operation.go

```diff
@@ -315,3 +315,105 @@ func Decode(serializedOperation string) (*longrunning.Operation, error) {
 	}
 	return op, nil
 }
+
+// ReliableStream implements a reliable operation stream.
+//
+// It keeps track of the operation name internally, and provides a Recv() func
+// which re-establishes the stream transparently if the operation name has been
+// established.
+type ReliableStream struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	client repb.ExecutionClient
+	stream repb.Execution_ExecuteClient
+	name   string
+}
+
+func NewReliableStream(ctx context.Context, client repb.ExecutionClient, stream repb.Execution_ExecuteClient, name string) *ReliableStream {
+	ctx, cancel := context.WithCancel(ctx)
+	return &ReliableStream{
+		ctx:    ctx,
+		cancel: cancel,
+		client: client,
+		stream: stream,
+		name:   name,
+	}
+}
+
+// Name returns the operation name, if known.
+func (s *ReliableStream) Name() string {
+	return s.name
+}
+
+// Recv attempts to reliably return the next operation on the named stream.
+//
+// If the stream is disconnected and the operation name has been received, it
+// will attempt to reconnect with WaitExecution.
+func (s *ReliableStream) Recv() (*longrunning.Operation, error) {
+	r := retry.DefaultWithContext(s.ctx)
+	for {
+		op, err := s.stream.Recv()
+		if err == nil {
+			if op.GetName() != "" {
+				s.name = op.GetName()
+			}
+			return op, nil
+		}
+		if !status.IsUnavailableError(err) || s.name == "" {
+			return nil, err
+		}
+		if !r.Next() {
+			return nil, s.ctx.Err()
+		}
+		req := &repb.WaitExecutionRequest{Name: s.name}
+		next, err := s.client.WaitExecution(s.ctx, req)
+		if err != nil {
+			return nil, err
+		}
+		s.stream.CloseSend()
+		s.stream = next
+	}
+}
+
+func (s *ReliableStream) Close() error {
+	if s.stream != nil {
+		s.stream.CloseSend()
+	}
+	s.stream = nil
+	s.client = nil
+	s.cancel()
+	return nil
+}
+
+// Message contains an operation along with its payload conveniently
+// unmarshaled.
+type Message struct {
```

Comment:
can we call this something else? it's easily confused with proto.Message

Reply:
renamed to ExecuteOperation in the other PR